tfhe/src/shortint/engine/server_side/shift.rs

```diff
@@ -47,14 +47,15 @@ impl ShortintEngine {
         Ok(())
     }
 
-    // by convention smart operations take mut refs to their inputs, even if they do not modify them
-    #[allow(clippy::needless_pass_by_ref_mut)]
     pub(crate) fn smart_scalar_left_shift(
         &mut self,
         server_key: &ServerKey,
         ct: &mut Ciphertext,
         shift: u8,
     ) -> EngineResult<Ciphertext> {
+        if !server_key.is_scalar_left_shift_possible(ct, shift) {
+            self.message_extract_assign(server_key, ct)?;
+        }
         let mut result = ct.clone();
         self.smart_scalar_left_shift_assign(server_key, &mut result, shift)?;
         Ok(result)
```

Comment:
I think this is possibly a bit sub optimal here.

Image a shift value that makes `is_scalar_left_shift` possible return false.
We do a message_extract_assign.

Then we call the `assign` version. however, inside the assign version the same check `is_scalar_left_shift_possible` may also return false because because maybe the shift value is too big which will in turn force the shift to be done with a PBS.

So we did 2 PBS, where we could only do one

Reply:
I guess the best way would be to `message_extract_assign` if it would make `scalar_left_shift` possible
But it's probably not worth the effort
Should I remove the whole commit then ?

Reply:
Removed